docs/source/crud/hooks.rst

```diff
@@ -0,0 +1,102 @@
+Hooks
+=====
+
+Hooks allow executing custom code as part of processing a crud request. You can use this to validate data, call another custom api, place messages on queues and many other things.
+
+Enabling a hook
+---------------
+
+Define a method, and register it with PiccoloCRUD:
+
+.. code-block:: python
+
+    # app.py
+    from piccolo_api.crud.endpoints import PiccoloCRUD
+
+    from movies.tables import Movie
+
+    # set movie rating to 10 before saving
+    async def set_movie_rating_10(row: Movie):
+        row.rating = 10
+        return row
+
+    # set movie rating to 20 before saving
+    async def set_movie_rating_10(row: Movie):
+        row.rating = 20
+        return row
+
+    async def pre_delete(row_id):
+        pass
+
+    # Register one or multiple hooks
+    app = PiccoloCRUD(table=Movie, read_only=False, hooks=[
+        Hook(hook_type=HookType.pre_save, coro=set_movie_rating_10),
+        Hook(hook_type=HookType.pre_save, coro=set_movie_rating_20),
+        Hook(hook_type=HookType.pre_delete, coro=pre_delete)
+        ]
+    )
+
+You can specify multiple hooks (also per hook_type). Hooks are executed in order. 
+You can use either async or regular functions.
+
+Hook types
+---------------
+
+There are different hook types, and each type takes a slightly different set of inputs. 
+It's also important to return the expected data from your hook.
+
+pre_save
+~~~~~~~~
+
+This hook runs during POST requests, prior to inserting data into the database.
+It takes a single parameter, ``row``, and should return the same:
+
+.. code-block:: python
+
+    async def set_movie_rating_10(row: Movie):
+        row.rating = 10
+        return row
+
+
+    app = PiccoloCRUD(table=Movie, read_only=False, hooks=[
+        Hook(hook_type=HookType.pre_save, coro=set_movie_rating_10)
+        ]
+    )
+
+pre_patch
+~~~~~~~~~
+
+This hook runs during PATCH requests, prior to changing the specified row in the database.
+It takes two parameters, ``row_id`` which is the id of the row to be changed, and ``values`` which is a dictionary of incoming values.
+Each function must return a dictionary which represent the data to be modified.
+
+.. code-block:: python
+
+    async def reset_name(row_id: int, values: dict):
+        current_db_row = await Movie.objects().get(Movie.id==row_id).run()
+        if values.get("name"):
+            values["name"] = values["name"].replace(" ", "")
+        return values
+
+    app = PiccoloCRUD(table=Movie, read_only=False, hooks=[
+        Hook(hook_type=HookType.pre_patch, coro=reset_name)
+        ]
+    )
+
+
+pre_delete
+~~~~~~~~~
```

Comment:
RST is super fussy - the number of characters in the underline need to match the heading. It's missing one tilde :p